src/datanode/src/heartbeat/handler/open_region.rs

```diff
@@ -0,0 +1,150 @@
+}
+
+impl HeartbeatResponseHandler for OpenTableHandler {
+    fn is_acceptable(&self, ctx: &HeartbeatResponseHandlerContext) -> bool {
+        matches!(
+            ctx.incoming_message.as_ref(),
+            Some((_, Instruction::OpenRegion { .. }))
+        )
+    }
+
+    fn handle(&self, ctx: &mut HeartbeatResponseHandlerContext) -> Result<()> {
+        match ctx.incoming_message.take() {
+            Some((
+                meta,
+                Instruction::OpenRegion {
+                    catalog,
+                    schema,
+                    table,
+                    table_id,
+                    engine,
+                    region_number,
+                },
+            )) => {
+                ctx.finish();
+
+                let mailbox = ctx.mailbox.clone();
+                let self_ref = Arc::new(self.clone());
+
+                common_runtime::spawn_bg(async move {
+                    let result = self_ref
+                        .open_region_inner(
+                            engine,
+                            OpenTableRequest {
+                                catalog_name: catalog,
+                                schema_name: schema,
+                                table_name: table,
+                                table_id,
+                                region_numbers: vec![region_number],
+                            },
+                        )
+                        .await;
+
+                    if let Err(e) = mailbox
+                        .send((
+                            meta,
+                            result.map_or_else(
+                                |error| InstructionReply::OpenRegion {
+                                    result: false,
+                                    error: Some(error.to_string()),
+                                },
+                                |result| InstructionReply::OpenRegion {
+                                    result,
+                                    error: None,
+                                },
+                            ),
+                        ))
+                        .await
+                    {
+                        error!(e;"Failed to send reply to mailbox");
+                    }
+                });
+            }
+            _ => warn!("Missing match in OpenTableHandler"),
+        }
+
+        Ok(())
+    }
+}
+
+impl OpenTableHandler {
+    async fn open_region_inner(&self, engine: String, request: OpenTableRequest) -> Result<bool> {
+        let OpenTableRequest {
+            catalog_name,
+            schema_name,
+            table_name,
+            ..
+        } = &request;
+        let engine =
+            self.table_engine_manager
+                .engine(&engine)
+                .context(error::TableEngineNotFoundSnafu {
+                    engine_name: &engine,
+                })?;
+        let ctx = EngineContext::default();
+
+        if let Some(table) = engine
```

Comment:
suggestion: I think we should get the table from the catalog manager first to see whether we have already opened it. If so, we can do nothing.

More specifically, `open_region_inner` should be idempotent.